### PR TITLE
[SPARK-21766][PySpark][SQL] DataFrame toPandas() raises ValueError with nullable int columns

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

When calling `DataFrame.toPandas()` (without Arrow enabled), if there is a `IntegralType` column (`IntegerType`, `ShortType`, `ByteType`) that has null values the following exception is thrown:

    ValueError: Cannot convert non-finite values (NA or inf) to integer

This is because the null values first get converted to float NaN during the construction of the Pandas DataFrame in `from_records`, and then it is attempted to be converted back to to an integer where it fails.

The fix is going to check if the Pandas DataFrame can cause such failure when converting, if so, we don't do the conversion and use the inferred type by Pandas.

Closes #18945

## How was this patch tested?

Added pyspark test.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
I'd go with this PR / approach. This approach and PR look pretty good. Let me help double check this tonight.

**Comment by viirya** [comment]
retest this please.

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -37,6 +37,7 @@
 from pyspark.sql.column import Column, _to_seq, _to_list, _to_java_column
 from pyspark.sql.readwriter import DataFrameWriter
 from pyspark.sql.streaming import DataStreamWriter
+from pyspark.sql.types import IntegralType
 from pyspark.sql.types import *
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: Looks we don't need to import `IntegralType`.

**Review by HyukjinKwon — APPROVED** [review]
LGTM. @viirya, are you online now? It'd be nicer if we add `Closes #18945` in the PR description.

File: python/pyspark/sql/dataframe.py [hunk 2]
```diff
@@ -1891,14 +1892,19 @@ def toPandas(self):
...
         else:
+            pdf = pd.DataFrame.from_records(self.collect(), columns=self.columns)
+
             dtype = {}
             for field in self.schema:
                 pandas_type = _to_corrected_pandas_type(field.dataType)
-                if pandas_type is not None:
+                # SPARK-21766: if an integer field is nullable and has null values, it can be
+                # inferred by pandas as float column. Once we convert the column with NaN back
+                # to integer type e.g., np.int16, we will hit exception.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: BTW, It'd be nicer if we add some comments saying like .. so, we use the inferred float type, not the corrected type from the schema.

File: python/pyspark/sql/dataframe.py (hunk 1 above)
Comment (viirya) [reply python/pyspark/sql/dataframe.py]: `IntegralType` is not in exposed list of `pyspark.sql.types`. Actually it is not used by any other places.

File: python/pyspark/sql/dataframe.py (hunk 2 above)
Comment (viirya) [reply python/pyspark/sql/dataframe.py]: Added. Thanks.

**Comment by viirya** [comment]
Add `Closes #18945` in the PR description.

File: python/pyspark/sql/dataframe.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/dataframe.py]: I see ...

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by viirya** [comment]
Thanks @HyukjinKwon

**Pull request closed without merging**